src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -0,0 +1,196 @@
+   *
+   * Please use `Topic` instead of `Queue` when all subscribers
+   * need to see each value enqueued.
+   *
+   * This process is equivalent to `dequeueBatch(1)`.
+   */
+  def dequeue: Stream[F, A]
+
+  /**
+    * Dequeue one `A` from this queue. Completes once one is ready.
+    */
+  def dequeue1: F[A]
+
+  /**
+   * The time-varying size of this Queue`. This signal refreshes
+   * only when size changes. Offsetting enqueues and de-queues may
+   * not result in refreshes.
+   */
+  def size: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * The size bound on the queue.
+   * Returns None if the queue is unbounded.
+   */
+  def upperBound: Option[Int]
+
+  /**
+   * Returns the available number of entries in the queue.
+   * Always returns `Int.MaxValue` when the queue is unbounded.
+   */
+  def available: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * Returns `true` when the queue has reached its upper size bound.
+   * Always returns `false` when the queue is unbounded.
+   */
+  def full: Stream[F,immutable.Signal[F,Boolean]]
+
+
+
+
+}
+
+object Queue {
+
+    private type DeQueue[F[_],A] = (A => F[Unit])
+
+
+  /**
+    * Internal state of the queue
+    * @param queue    Queue, expressed as vector for fast cons/uncons from head/tail
+    * @param deq      A list of De-Queuer (filled up when queue is empty)
+    * @tparam F
+    * @tparam A
+    */
+    private case class State[F[_],A](
+      queue: Vector[A]
+      , deq:Vector[DeQueue[F,A]]
+    )
+
+
+    def apply[F[_],A](bound:Int = 0)(implicit F:AsyncExt[F]):Stream[F,Queue[F,A]] = {
+      Signal(0).flatMap { szSignal =>
+        Stream.eval {
+          F.bind(F.ref[State[F,A]]) { qref =>
+            F.map(F.set(qref)(F.pure(State(Vector.empty,Vector.empty)))) { _ =>
+
+              // Signals size change of queue, if that has changed
+              def signalSize(s:State[F,A], ns:State[F,A]) : F[Unit] = {
+                if (s.queue.size != ns.queue.size) szSignal.set(ns.queue.size)
+                else F.pure(())
+              }
+
+              // Offers one element to queue
+              // if queue is full completes immediately with false
+              // if the queue is empty, and some deq are waiting, fills the head of deq.
+              def _offer1(a: A): F[Boolean] = {
+                F.bind(F.ref[Boolean]) { enqRef =>
+                  F.bind(F.modify(qref){ s =>
+                    if (bound >= 0 && s.queue.size -1 >= bound) F.map(F.setPure(enqRef)(false)){ _ => s}
+                    else {
+                      s.deq headOption match {
+                        case None =>
+                          F.map(F.setPure(enqRef)(true)){ _ => s.copy(queue = s.queue :+ a ) }
+                        case Some(deq) =>
+                          F.bind(deq(a)){_ =>
+                            F.map(F.setPure(enqRef)(true)) { _ => s }
+                          }
+                      }
+                    }
+                  }){ case Change(s,ns) =>
```

Comment:
Not a big deal, but it took me a minute to figure out what `s` and `ns` meant. (`ns` is abbreviation for 'new state' I'm guessing?) I might do `{ change => ... signalSize(change.previous, change.now) }` so the reader doesn't have to remember what `s` and `ns` are bound to.